### PR TITLE
Add some unit tests for network.py

### PR DESCRIPTION
**Pull request by vibss2397**
Created tests for some of the functions in network,py, do check and tell if any changes are required.

(Earlier discussion about this PR is in #93)

### CODE REVIEW

**Comment by vibss2397** [comment]
@marco-c i wanted to ask are there some other functions in this file which need to be tested?
Also, my uni exams are coming up so was a bit busy for the past few days

**Comment by marco-c** [comment]
> @marco-c i wanted to ask are there some other functions in this file which need to be tested?

Yes, we should try to test as many functions as possible. This PR is only adding tests for a subset of the functions.
You can do it in a follow-up PR.

File: tests/test_network.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+arr1 = np.array([[1, 1]], dtype=np.float32)
+arr2 = np.array([[0, 1]], dtype=np.float32)
+
+arr3 = np.array([[0, 0]], dtype=np.float32)
+arr4 = np.array([[3, 2]], dtype=np.float32)
+
+
+def test_eucledian_distance():
+    dist = network.euclidean_distance([arr1, arr2])
+    evaluate = K.eval(dist)
```
Comment (marco-c) [inline tests/test_network.py]: Remove the `evaluate` variable and just do the assert here

File: tests/test_network.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    dist2 = network.euclidean_distance([arr3, arr4])
+    evaluate2 = K.eval(dist2)
```
Comment (marco-c) [inline tests/test_network.py]: Same here

File: tests/test_network.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    assert (evaluate == 1)
+    assert (math.floor(evaluate2 * 100) == 360)
+
+
+def test_eucl_distance_output_shape():
+    vect = [arr1.shape, arr2.shape]
+    shape = network.eucl_dist_output_shape(vect)
+    assert (shape == (1, 1))
```
Comment (marco-c) [inline tests/test_network.py]: Don't create `vect` and `shape` variables, just assert the call directly like this:
```Python
assert(network.eucl_dist_output_shape([arr1.shape, arr2.shape]) == (1, 1))
```

Also test arr4

File: tests/test_network.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+
+def test_contrastive_loss():
```
Comment (marco-c) [inline tests/test_network.py]: This seems overly complex. As a unit test, this should just test the `contrastive_loss` function.

**Comment by vibss2397** [comment]
@marco-c for the models in the network file, do we test the shape of inputs, outputs or they have weights or what do we want to test?

File: tests/test_network.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    assert (math.floor(K.eval(dist2) * 100) == 360)
+
+
+def test_eucl_distance_output_shape():
+    vect = [arr1.shape, arr2.shape]
+    assert (network.eucl_dist_output_shape(vect) == (1, 1))
+
+
+def test_contrastive_loss():
+    label = [0, 1]  # possible values for output
```
Comment (marco-c) [inline tests/test_network.py]: Why defining an array here? Just use the raw values when needed.

File: tests/test_network.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    euclid_dist = network.euclidean_distance([arr1, arr2])  # it returns 1
```
Comment (marco-c) [inline tests/test_network.py]: You're testing `contrastive_loss` here, so there's no pointing in calling `euclidean_distance` if you already know its result.

File: tests/test_network.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+def test_eucledian_distance():
+    dist = network.euclidean_distance([arr1, arr2])
+    dist2 = network.euclidean_distance([arr3, arr4])
+    assert (K.eval(dist) == 1)
+    assert (math.floor(K.eval(dist2) * 100) == 360)
+
+
+def test_eucl_distance_output_shape():
+    vect = [arr1.shape, arr2.shape]
```
Comment (marco-c) [inline tests/test_network.py]: Already told in my earlier comment, please remove the `vect` variable.

Comment (marco-c) [reply tests/test_network.py]: Also, as I said before, also test arr4

File: tests/test_network.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+arr2 = np.array([[0, 1]], dtype=np.float32)
+
+arr3 = np.array([[0, 0]], dtype=np.float32)
+arr4 = np.array([[3, 2]], dtype=np.float32)
+
+
+def test_eucledian_distance():
+    dist = network.euclidean_distance([arr1, arr2])
+    dist2 = network.euclidean_distance([arr3, arr4])
+    assert (K.eval(dist) == 1)
```
Comment (marco-c) [inline tests/test_network.py]: Nit: this is a weird way to arrange the assertions. Put the assertion related to a `euclidean_distance` call right after the call itself.

File: tests/test_network.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+arr1 = np.array([[1, 1]], dtype=np.float32)
+arr2 = np.array([[0, 1]], dtype=np.float32)
+
+arr3 = np.array([[0, 0]], dtype=np.float32)
+arr4 = np.array([[3, 2]], dtype=np.float32)
+
+
+def test_eucledian_distance():
+    dist = network.euclidean_distance([arr1, arr2])
+    dist2 = network.euclidean_distance([arr3, arr4])
```
Comment (marco-c) [inline tests/test_network.py]: Use `dist` and not `dist2` when you move the assertions

File: tests/test_network.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    loss1 = network.contrastive_loss(label[0], euclid_dist)
+    loss2 = network.contrastive_loss(label[1], euclid_dist)
+    assert (K.eval(loss1) == 0)  # since the output and the label are far away
+    assert (K.eval(loss2) == 1)  # since the output and the label are close
+
+
+def test_accuracy():
+    y_pred = np.array([0, 0.4, 0.7, 0.6, 0.5])
+    y_true = np.array([0, 1, 0, 0, 1])
+    assert (math.floor(K.eval(network.accuracy(y_true, y_pred)) * 100) == 60)
```
Comment (marco-c) [inline tests/test_network.py]: Nit: no space between assert and `(`, here and in the rest of the file.

**Comment by vibss2397** [comment]
@marco-c done with the changes

File: tests/test_network.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+
+def test_eucledian_distance():
+    dist = network.euclidean_distance([arr1, arr2])
+    assert (K.eval(dist) == 1)
+
+
+def test_eucl_distance_output_shape():
+    assert(network.eucl_dist_output_shape([arr1.shape, arr2.shape]) == (1, 1))
+    assert(network.eucl_dist_output_shape([arr3.shape, arr4.shape]) == (1, 1))
```
Comment (marco-c) [inline tests/test_network.py]: Can you add a test with another array that has a different shape? Just to have a different output than just (1, 1).

**Pull request merged**